src/client/game.cpp

```diff
@@ -4257,6 +4218,54 @@ void Game::updateShadows()
 	shadow->getDirectionalLight().update_frustum(camera, client, m_camera_offset_changed);
 }
 
+void Game::drawScene(ProfilerGraph *graph, RunStats *stats, f32 dtime)
+{
+	const video::SColor skycolor = this->sky->getSkyColor();
+
+	TimeTaker tt_draw("Draw scene", nullptr, PRECISION_MICRO);
+	this->driver->beginScene(true, true, skycolor);
+
+	const LocalPlayer *player = client->getEnv().getLocalPlayer();
+	bool draw_wield_tool = (this->m_game_ui->m_flags.show_hud &&
+			(player->hud_flags & HUD_FLAG_WIELDITEM_VISIBLE) &&
+			(this->camera->getCameraMode() == CAMERA_MODE_FIRST));
+	bool draw_crosshair = (
+			(player->hud_flags & HUD_FLAG_CROSSHAIR_VISIBLE) &&
+			(this->camera->getCameraMode() != CAMERA_MODE_THIRD_FRONT));
+#ifdef HAVE_TOUCHSCREENGUI
+	if (this->isNoCrosshairAllowed())
+		draw_crosshair = false;
+#endif
+	this->m_rendering_engine->draw_scene(skycolor, this->m_game_ui->m_flags.show_hud,
+			this->m_game_ui->m_flags.show_minimap, draw_wield_tool, draw_crosshair);
+
+	/*
+		Profiler graph
+	*/
+	v2u32 screensize = this->driver->getScreenSize();
+
+	if (this->m_game_ui->m_flags.show_profiler_graph)
+		graph->draw(10, screensize.Y - 10, driver, g_fontengine->getFont());
+
+	/*
+		Damage flash
+	*/
+	if (runData.damage_flash > 0.0f) {
+		video::SColor color(runData.damage_flash, 180, 0, 0);
+		this->driver->draw2DRectangle(color,
+					core::rect<s32>(0, 0, screensize.X, screensize.Y),
+					NULL);
+
+		runData.damage_flash -= 384.0f * dtime;
```

Comment:
Can you move this single line back to `updateFrame`?  Then you could remove the `dtime` argument from `drawScene`, and all state keeping would be gone from `drawScene`. This would help with #14058.